src/data/mutable_data.rs

```diff
@@ -308,3 +318,236 @@ impl Debug for MutableData {
+        let mut data = BTreeMap::new();
+        for i in 0..105 {
+            let _ = data.insert(vec![i], val.clone());
+        }
+        assert!(MutableData::new(rand::random(), 0, BTreeMap::new(), data, BTreeSet::new())
+            .is_err());
+
+        let mut data = BTreeMap::new();
+        for i in 0..100 {
+            let _ = data.insert(vec![i], val.clone());
+        }
+
+        let (owner, _) = sign::gen_keypair();
+
+        let mut owners = BTreeSet::new();
+        assert!(owners.insert(owner), true);
+
+        let mut md = unwrap!(MutableData::new(rand::random(), 0, BTreeMap::new(), data, owners));
+
+        assert_eq!(md.keys().len(), 100);
+        assert_eq!(md.values().len(), 100);
+        assert_eq!(md.entries().len(), 100);
+
+        // Try to get over the limit
+        assert!(md.ins_entry(vec![101u8], val.clone(), owner).is_ok());
+        assert!(md.ins_entry(vec![102u8], val.clone(), owner).is_err());
+
+        assert!(md.del_entry(&vec![101u8], owner).is_ok());
+        assert!(md.ins_entry(vec![102u8], val.clone(), owner).is_ok());
+    }
+
+    #[test]
+    fn size_limit() {
+        let big_val = Value {
+            content: iter::repeat(0)
+                .take((MAX_MUTABLE_DATA_SIZE_IN_BYTES - 1024) as usize)
+                .collect(),
+            entry_version: 0,
+        };
+
+        let small_val = Value {
+            content: iter::repeat(0).take(2048).collect(),
+            entry_version: 0,
+        };
+
+        // It must not be possible to create MutableData with size of more than 1 MiB
+        let mut data = BTreeMap::new();
+        let _ = data.insert(vec![0], big_val.clone());
+        let _ = data.insert(vec![1], small_val.clone());
+
+        assert!(MutableData::new(rand::random(), 0, BTreeMap::new(), data, BTreeSet::new())
+            .is_err());
+
+        let mut data = BTreeMap::new();
+        let _ = data.insert(vec![0], big_val.clone());
+
+        let (owner, _) = sign::gen_keypair();
+        let mut owners = BTreeSet::new();
+        assert!(owners.insert(owner), true);
+
+        let mut md = unwrap!(MutableData::new(rand::random(), 0, BTreeMap::new(), data, owners));
+
+        // Try to get over the mutation limit of 2 MiB
+        assert!(md.ins_entry(vec![1], big_val.clone(), owner).is_ok());
+        assert!(md.ins_entry(vec![2], small_val.clone(), owner).is_err());
+        assert!(md.del_entry(&vec![0], owner).is_ok());
+        assert!(md.ins_entry(vec![0], small_val, owner).is_ok());
+    }
+
+    #[test]
+    fn transfer_ownership() {
+        let (owner, _) = sign::gen_keypair();
+        let (pk1, _) = sign::gen_keypair();
+
+        let mut owners = BTreeSet::new();
+        owners.insert(owner);
+
+        let mut md =
+            unwrap!(MutableData::new(rand::random(), 0, BTreeMap::new(), BTreeMap::new(), owners));
+
+        // Try to do ownership transfer from a non-owner requester
+        assert!(md.change_owner(pk1, pk1).is_err());
```

Comment:
It is better to check for specific error, that way frontend, safe_core, routing and vault always agree on the same error (unlike right now where safe_core expects NoSuchAccount and vault (incorrectly) returns NoSuchData).